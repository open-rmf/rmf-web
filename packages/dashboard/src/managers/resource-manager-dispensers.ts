interface Location {
  x: number;
  y: number;
  yaw: number;
  level_name: string;
}

export interface RawDispenserResource {
  icons: Record<string, string>;
  location: Location;
}

export interface DispenserResource extends RawDispenserResource {
  guid: string;
}

export class DispenserResourceManager {
  dispensers: Record<string, DispenserResource>;

  constructor(dispenserResources: Record<string, RawDispenserResource>) {
    this.dispensers = this.assignGuidToDispensers(dispenserResources);
  }

  getIconPath = (dispenserName: string): string | null => {
    if (!this.dispenserExists(dispenserName)) {
      return null;
    }
    if (!this.dispensers[dispenserName].hasOwnProperty('icons')) {
      return null;
    }
    const rootIconPath = '/assets/icons';
    const dispenserIcon = this.dispensers[dispenserName].icons[dispenserName];

    return dispenserIcon ? `${rootIconPath}${dispenserIcon}` : null;
  };

  get all(): Record<string, DispenserResource> {
    return this.dispensers;
  }

  get allValues(): DispenserResource[] {
    return Object.values(this.dispensers);
  }

  private assignGuidToDispensers(
    dispensers: Record<string, RawDispenserResource>,
  ): Record<string, DispenserResource> {
    let newDict: any = Object.assign({}, dispensers);
    Object.keys(dispensers).forEach((key) => {
      newDict[key].guid = key;
    });
    return newDict as Record<string, DispenserResource>;
  }

  private dispenserExists = (dispenserName: string) => {
    if (!this.dispensers.hasOwnProperty(dispenserName)) {
      return false;
    }
    return true;
  };
}
