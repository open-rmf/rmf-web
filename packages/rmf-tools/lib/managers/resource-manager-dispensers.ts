import Debug from 'debug';

const debug = Debug('ResourceManager');

interface Location {
  x: number;
  y: number;
  yaw: number;
  level_name: string;
}

export interface RawDispenserResource {
  icons: Record<string, string>;
  location: Location;
}

export interface DispenserResource extends RawDispenserResource {
  guid: string;
}

export class DispenserResourceManager {
  dispensers: Record<string, DispenserResource>;

  constructor(dispenserResources: Record<string, RawDispenserResource>) {
    this.dispensers = this.assignGuidToDispensers(dispenserResources);
  }

  getIconPath = async (dispenserName: string): Promise<string | null> => {
    if (!this.dispenserExists(dispenserName)) {
      debug(`failed to load icon for "${dispenserName}" (dispenser does not exist in resources)`);
      return null;
    }
    if (!Object.prototype.hasOwnProperty.call(this.dispensers[dispenserName], 'icons')) {
      debug(`failed to load icon for "${dispenserName}" (dispenser does not have an icon)`);
      return null;
    }
    const dispenserIcon = this.dispensers[dispenserName].icons[dispenserName];

    const currDir = process.cwd();

    try {
      return (
        await import(/* webpackMode: "eager" */ `${currDir}/src/assets/resources${dispenserIcon}`)
      ).default;
    } catch {
      debug(`failed to load icon for "${dispenserName}" (failed to load icon module)`);
      return null;
    }
  };

  get all(): Record<string, DispenserResource> {
    return this.dispensers;
  }

  get allValues(): DispenserResource[] {
    return Object.values(this.dispensers);
  }

  private assignGuidToDispensers(
    dispensers: Record<string, RawDispenserResource>,
  ): Record<string, DispenserResource> {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const newDict: any = Object.assign({}, dispensers);
    Object.keys(dispensers).forEach((key) => {
      newDict[key].guid = key;
    });
    return newDict as Record<string, DispenserResource>;
  }

  private dispenserExists = (dispenserName: string) => {
    if (!Object.prototype.hasOwnProperty.call(this.dispensers, dispenserName)) {
      return false;
    }
    return true;
  };
}
