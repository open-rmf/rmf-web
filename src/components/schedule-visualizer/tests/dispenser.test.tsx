import { createMount } from '@material-ui/core/test-utils';
import React from 'react';
import fakeResources from '../../../mock/data/resources';
import Dispenser from '../dispenser';
import fakeDispenserStates from '../../../mock/data/dispenser-states';
import toJson from 'enzyme-to-json';
import { DispenserResourceManager } from '../../../resource-manager-dispensers';

const mount = createMount();

describe('Dispenser component', () => {
  let dispenserHandler: DispenserResourceManager;

  beforeEach(() => {
    const dispenserResources = fakeResources().dispensers;
    if (!dispenserResources || !Object.keys(dispenserResources).length) {
      throw new Error('To execute this test you need to instantiate a dispenser');
    }
    dispenserHandler = new DispenserResourceManager(dispenserResources);
  });

  test('Fires click event when it has a state', () => {
    let clicked = false;

    const dispenserStates = fakeDispenserStates();
    const dispensers = dispenserHandler.allValues;
    const root = mount(
      <svg>
        <Dispenser
          dispenser={dispensers[0]}
          dispenserHandler={dispenserHandler}
          footprint={0.5}
          onClick={() => (clicked = true)}
          dispenserState={dispenserStates[dispensers[0].guid]}
        />
      </svg>,
    );

    root.find(Dispenser).simulate('click');
    expect(clicked).toBe(true);

    root.unmount();
  });

  test(`Don't Fire click event when it has no state`, () => {
    /**
     * We may want to add dispensers to the resource file, but we may not have a way to
     * know the status of that dispenser or the status may not be sent by the backend
     * because it is a new dispenser not yet registered. This means that if the dispenser
     * icon will be shown on the map but it will not have a status in the omnipanel.
     * Therefore, if a click is pressed on the dispenser, it will set an undefined state
     * in the omnipanel.
     */
    let clicked = false;
    const dispensers = dispenserHandler.allValues;
    const root = mount(
      <svg>
        <Dispenser
          dispenser={dispensers[0]}
          footprint={0.5}
          onClick={() => (clicked = true)}
          dispenserHandler={dispenserHandler}
        />
      </svg>,
    );

    root.find(Dispenser).simulate('click');
    expect(clicked).toBe(false);

    root.unmount();
  });

  test(`Renders correctly`, () => {
    const dispensers = dispenserHandler.allValues;
    const wrapper = mount(
      <svg>
        <Dispenser dispenser={dispensers[0]} footprint={0.5} dispenserHandler={dispenserHandler} />
      </svg>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
