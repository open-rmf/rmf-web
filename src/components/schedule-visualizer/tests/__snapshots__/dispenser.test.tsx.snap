// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dispenser component Renders correctly 1`] = `
<svg>
  <ForwardRef
    dispenser={
      Object {
        "guid": "coke_ingestor",
        "icons": Object {
          "coke_ingestor": "",
        },
        "location": Object {
          "level_name": "L1",
          "x": 20.95,
          "y": -6.95,
          "yaw": 0,
        },
      }
    }
    dispenserHandler={
      DispenserResourceManager {
        "dispenserExists": [Function],
        "dispensers": Object {
          "coke_dispenser": Object {
            "guid": "coke_dispenser",
            "icons": Object {
              "coke_dispenser": "",
            },
            "location": Object {
              "level_name": "L1",
              "x": 16.85,
              "y": -4.8,
              "yaw": 0,
            },
          },
          "coke_ingestor": Object {
            "guid": "coke_ingestor",
            "icons": Object {
              "coke_ingestor": "",
            },
            "location": Object {
              "level_name": "L1",
              "x": 20.95,
              "y": -6.95,
              "yaw": 0,
            },
          },
        },
        "getIconPath": [Function],
      }
    }
    footprint={0.5}
  >
    <g
      className="makeStyles-container-8"
      data-component="Dispenser"
      data-name="coke_ingestor"
      data-state="false"
      onClick={[Function]}
      transform="translate(20.95 6.95)
        rotate(0)"
    >
      <DispenserDefaultIcon
        footprint={0.5}
      >
        <svg
          height={1}
          viewBox="0 0 24 24"
          width={1}
          x={-0.5}
          xmlns="http://www.w3.org/2000/svg"
          y={-0.5}
        >
          <path
            className="makeStyles-clickable-9"
            d="M19 3H4.99c-1.11 0-1.98.9-1.98 2L3 19c0 1.1.88 2 1.99 2H19c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 12h-4c0 1.66-1.35 3-3 3s-3-1.34-3-3H4.99V5H19v10zm-3-5h-2V7h-4v3H8l4 4 4-4z"
            fill="#4caf50"
          />
        </svg>
      </DispenserDefaultIcon>
    </g>
    <SvgText
      className="makeStyles-dispenserText-7"
      id="dispenserName"
      targetWidth={1.1}
      text="coke_ingestor"
      x={20.95}
      y={6.95}
    >
      <text
        className="makeStyles-dispenserText-7"
        id="dispenserName"
        x={20.95}
        y={6.95}
      />
    </SvgText>
  </ForwardRef>
</svg>
`;
